pvlib/horizon.py

```diff
@@ -0,0 +1,646 @@
+    """
+    Calculates a horizon profile from a three grids containing lat, lon,
+    and elevation values. The "site" is assumed to be at the center of the
+    grid.
+
+    Parameters
+    ----------
+    lat_grid : ndarray
+        A 2d array containing latitude values that correspond to the other
+        two input grids.
+
+    lon_grid : ndarray
+        A 2d array containing longitude values that correspond to the other
+        two input grids.
+
+    elev_grid : ndarray
+        A 2d array containing elevation values that correspond to the other
+        two input grids.
+
+    sampling_method : string
+        A string that specifies the sampling method used to generate the
+        horizon profile.
+
+    sampling_param : variable
+        A parameter that is passed into the function specified by
+        sampling_method.
+
+    Returns
+    -------
+    bearing_deg: Nx1 ndarray
+        The bearings from the "site" to sampled points in degrees
+        East of North.
+
+    elevation_angle_deg: numeric
+        The angles that the sampled points make with the horizontal
+        as observed from the "site". Given in degrees above the
+        horizontal.
+
+    """
+    assert(lat_grid.shape == lon_grid.shape == elev_grid.shape)
+
+    grid_shape = lat_grid.shape
+    grid_center_i = (grid_shape[0] - 1) // 2
+    grid_center_j = (grid_shape[1] - 1) // 2
+    site_lat = lat_grid[grid_center_i, grid_center_j]
+    site_lon = lon_grid[grid_center_i, grid_center_j]
+    site_elev = elev_grid[grid_center_i, grid_center_j]
+    site = np.array([site_lat, site_lon, site_elev])
+
+    if sampling_method == "grid":
+        samples = sample_using_grid(lat_grid, lon_grid, elev_grid)
+    elif sampling_method == "triangles":
+        samples = sample_using_triangles(lat_grid, lon_grid, elev_grid,
+                                         sampling_param)
+    elif sampling_method == "interpolator":
+        samples = sample_using_interpolator(lat_grid, lon_grid, elev_grid,
+                                            sampling_param)
+
+    bearing_deg, elevation_angle_deg = elevation_angle_calc(site, samples)
+
+    return bearing_deg, elevation_angle_deg
+
+
+def sample_using_grid(lat_grid, lon_grid, elev_grid):
+    """
+    Calculates the Elevation angle from the site (center of the grid)
+    to every point on the grid.
+
+    Parameters
+    ----------
+    lat_grid : ndarray
+        A 2d array containing latitude values that correspond to the other
+        two input grids.
+
+    lon_grid : ndarray
+        A 2d array containing longitude values that correspond to the other
+        two input grids.
+
+    elev_grid : ndarray
+        A 2d array containing elevation values that correspond to the other
+        two input grids.
+
+
+    Returns
+    -------
+    all_samples: Nx3 ndarray
+        Array of lat, lon, elev points that are grid points.
+    """
+    assert(lat_grid.shape == lon_grid.shape == elev_grid.shape)
```

Comment:
I don't think we need the `assert` here and in the other `sample` functions, since grid size equality is tested in `calculate_horizon_points`